backend/de.metas.swat/de.metas.swat.base/src/main/java/org/adempiere/invoice/process/C_Invoice_GenerateCorrectionInvoice.java

```diff
@@ -0,0 +1,153 @@
+
+	private static final DocBaseAndSubType SALES_INVOICE = DocBaseAndSubType.of(DOCBASETYPE_ARInvoice);
+	private static final DocBaseAndSubType CORRECTION_INVOICE = DocBaseAndSubType.of(X_C_DocType.DOCBASETYPE_ARInvoice, X_C_DocType.DOCSUBTYPE_CorrectionInvoice);
+	private static final int C_DocType_ID = -1;
+	private static final AdMessageKey MSG_Event_DocumentGenerated = AdMessageKey.of("Event_DocumentGenerated");
+
+	@Override
+	public ProcessPreconditionsResolution checkPreconditionsApplicable(final @NonNull IProcessPreconditionsContext context)
+	{
+		if (!context.isSingleSelection())
+		{
+			return ProcessPreconditionsResolution.rejectBecauseNotSingleSelection().toInternal();
+		}
+
+		final InvoiceId invoiceId = InvoiceId.ofRepoId(context.getSingleSelectedRecordId());
+		final I_C_Invoice invoiceRecord = invoiceDAO.getByIdInTrx(invoiceId);
+		final DocTypeId docTypeId = DocTypeId.ofRepoId(invoiceRecord.getC_DocType_ID());
+		final DocBaseAndSubType docBaseAndSubType = docTypeDAO.getDocBaseAndSubTypeById(docTypeId);
+
+		if(!(docBaseAndSubType.equals(SALES_INVOICE) || invoiceBL.isAdjustmentCharge(invoiceRecord)))
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("Not an basic Sales Invoice or Adjustment Charge");
+		}
+
+		final DocStatus docStatus = DocStatus.ofCode(invoiceRecord.getDocStatus());
+		if(!docStatus.isCompletedOrClosed())
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("Not completed or closed");
+		}
+
+		if(!countryDAO.isEnforceCorrectionInvoice(invoiceBL.getBillToCountryId(invoiceId)))
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("Not isEnforceCorrectionInvoice");
+		}
+
+		return ProcessPreconditionsResolution.accept();
+	}
+
+	@Override
+	protected String doIt() throws Exception
+	{
+		final de.metas.adempiere.model.I_C_Invoice invoice = InterfaceWrapperHelper.create(getCtx(), getRecord_ID(), de.metas.adempiere.model.I_C_Invoice.class, get_TrxName());
+
+		final InvoiceCreditContext creditCtx = InvoiceCreditContext.builder()
+				.docTypeId(null)
+				.completeAndAllocate(true)
+				.referenceOriginalOrder(true)
+				.referenceInvoice(true)
+				.creditedInvoiceReinvoicable(false)
+				.fixedInvoice(true)
+				.build();
+
+		final I_C_Invoice creditMemo = invoiceBL.creditInvoice(invoice, creditCtx);
+		createUserNotification(creditMemo);
+
+		final AdjustmentChargeCreateRequest adjustmentChargeCreateRequest = AdjustmentChargeCreateRequest.builder()
+				.invoiceID(InvoiceId.ofRepoId(getRecord_ID()))
+				.docBaseAndSubTYpe(CORRECTION_INVOICE)
+				.build();
+
+		final I_C_Invoice correctionInvoice = invoiceBL.adjustmentCharge(adjustmentChargeCreateRequest);
+		createUserNotification(correctionInvoice);
+
+		return MSG_OK;
+	}
+
+	private void createUserNotification(@NonNull final I_C_Invoice invoiceRecord)
+	{
+		final TableRecordReference invoiceRef = TableRecordReference.of(invoiceRecord);
```

Comment:
why not using org.adempiere.invoice.event.InvoiceUserNotificationsProducer?

Reply:
1. i didn't know it exists
2. it doesn't support adjustment to docType changes like my implementation does (docTypeName fix in AD_Message)

Reply:
how ur message looks like? (example)

I have the feeling we can improve the InvoiceUserNotificationsProducer and use that format...

Reply:
![image](https://user-images.githubusercontent.com/32903793/224124647-777fbf08-d722-44fe-b3f1-679a92609017.png)
The thing is the current trl of the used doctype is used and the ad_msg doesn't need to be adjusted